### PR TITLE
[REF] Renderer: convert RendererPlugin to store 

### PR DESCRIPTION
**Pull request by rrahir**
Move the main contributor to draw on the grid as a dedicated store.

This revision is part of a wider goal to remove all drawing from Model.

This one was tricky, the access of data through reactivity is really
slow and that slowness  coupled with a forced rendering at each model
update makes the sheet unusably slow (400 ms of rendering time).

Task: 3719197

## Description:

description of this task, what is implemented and why it is implemented that way.

Task: : [TASK_ID](https://www.odoo.com/web#id=TASK_ID&action=333&active_id=2328&model=project.task&view_type=form&cids=1&menu_id=4720)

## review checklist

- [ ] feature is organized in plugin, or UI components
- [ ] support of duplicate sheet (deep copy)
- [ ] in model/core: ranges are Range object, and can be adapted (adaptRanges)
- [ ] in model/UI: ranges are strings (to show the user)
- [ ] undo-able commands (uses this.history.update)
- [ ] multiuser-able commands (has inverse commands and transformations where needed)
- [ ] new/updated/removed commands are documented
- [ ] exportable in excel
- [ ] translations (\_t("qmsdf %s", abc))
- [ ] unit tested
- [ ] clean commented code
- [ ] track breaking changes
- [ ] doc is rebuild (npm run doc)
- [ ] status is correct in Odoo

### CODE REVIEW

**Comment by rrahir** [comment]
missing a mock in the tests (we take 25seconds more to run the suit right now)

**Comment by rrahir** [comment]
> missing a mock in the tests (we take 25seconds more to run the suit right now)

Mock added, the test time looks prettymuch unchanged

File: src/stores/renderer_store.ts [hunk 1]
```diff
@@ -1,14 +1,11 @@
...
-import { ModelStore } from "./model_store";
 
 interface Renderer {
   drawLayer(ctx: GridRenderingContext, layer: LAYERS): void;
   renderingLayers: LAYERS[];
 }
 
-export class RendererStore extends ReactiveStore {
-  protected model = this.get(ModelStore);
+export class RendererStore {
```
Comment (rrahir) [inline src/stores/renderer_store.ts:8]: does this have to be reactive? doens't feel like it in fine

File: src/stores/grid_renderer_store.ts [hunk 2]
```diff
@@ -33,72 +35,41 @@ import {
...
-    const isColHidden = this.getters.isColHidden(sheetId, col);
-    return {
-      start,
-      size: size,
-      end: start + (isColHidden ? 0 : size),
-    };
+export class GridRendererStore extends SpreadsheetStore {
+  constructor(get: Get) {
+    super(get);
+    markRaw(this);
```
Comment (rrahir) [inline src/stores/grid_renderer_store.ts:63]: not sure about this (if it should be here or jus tin the renderedstore) - if both renderedStore and this one are reactive, then the time spent in reactive becomes huge, like 20+ times slower...
Should we have different type of spreadsheet stores? reactive and not reactive?

File: src/store_engine/store.ts [hunk 3]
```diff
@@ -89,6 +89,7 @@ type NeverReturns<T> = T extends (...args: any[]) => any ? (...args: Parameters<
 export class ReactiveStore {
   constructor(protected get: Get) {
     return reactive(this);
+    // return this
```
Comment (LucasLefevre) [inline src/store_engine/store.ts:92]: :eyes:

File: src/store_engine/store.ts [hunk 4]
```diff
@@ -76,7 +76,7 @@ export type Store<T> = CQS<T>;
  * [2] https://en.wikipedia.org/wiki/Command%E2%80%93query_separation
  * [3] in an ideal world, they would be deeply read-only, but that's not possible natively in TypeScript
  */
-type CQS<T> = {
+export type CQS<T> = {
```
Comment (LucasLefevre) [inline src/store_engine/store.ts:79]: you can revert this and use `Store<T>` instead

File: src/stores/grid_renderer_store.ts [hunk 5]
```diff
@@ -1,4 +1,7 @@
-import { ICONS } from "../../components/icons/icons";
+// import { markRaw } from "@odoo/owl";
```
Comment (LucasLefevre) [inline src/stores/grid_renderer_store.ts:1]: :eyes:

File: tests/test_helpers/helpers.ts [hunk 6]
```diff
@@ -138,7 +140,13 @@ export function makeTestEnv(mockEnv: Partial<SpreadsheetChildEnv> = {}): Spreads
     raiseError: mockEnv.raiseError || (() => {}),
     askConfirmation: mockEnv.askConfirmation || (() => {}),
   };
+  // TODORAR : type case or create a fake class that does nothing with its drawlayer ?
+  const gridRendererStore = {
+    drawLayer: (renderingContext: GridRenderingContext, layer: LAYERS) => {},
+  } as GridRenderer;
+
   container.inject(NotificationStore, notificationStore);
+  container.inject(GridRenderer, gridRendererStore);
```
Comment (LucasLefevre) [inline tests/test_helpers/helpers.ts:149]: we could actually inject a mock instance of `RendererStore` to cancel all renderings, not only the grid renderer

File: src/stores/grid_renderer_store.ts [hunk 7]
```diff
@@ -33,70 +35,49 @@ import {
...
 export const CELL_BACKGROUND_GRIDLINE_STROKE_STYLE = "#111";
 
-export class RendererPlugin extends UIPlugin {
-  static layers = [LAYERS.Background, LAYERS.Headers];
-  static getters = ["getColDimensionsInViewport", "getRowDimensionsInViewport"] as const;
-
-  private boxes: Box[] = [];
+export class GridRenderer {
+  private getters!: Getters;
+  private renderer!: Store<RendererStore>;
```
Comment (LucasLefevre) [inline src/stores/grid_renderer_store.ts]:
```suggestion
  private getters: Getters;
  private renderer: Store<RendererStore>;
```

**Review by LucasLefevre — COMMENTED** [review]
robodoo r+ rebase-ff

**Pull request closed without merging**